Fix: Updated Proposal process

VitWW opened the pull request:
This is not a Proposal, but an update of description of Review Process in README file.

simonpj commented:
Good change. 

But it's still pretty hard to get an overview because of the detail.  [Here is a new attempt to make it clear](https://docs.google.com/document/d/1mKP5_ewNr5DanKCGvdjqC5SntNs8R5J-Wbho0-_eD0c/edit?usp=sharing), currently as a google doc. If people like it, mabye someone could transition it into MarkDown?

noughtmare commented:
[Rendered README](https://github.com/VitWW/ghc-proposals/blob/update-readme/README.rst)

simonpj commented:
Great!  let's see what others say, esp @adamgundry

VitWW commented:
Add After-accepted stage as @mpickering proposed.

adamgundry commented on README.rst:
```diff
@@ -80,16 +120,22 @@ This section outlines what stages a proposal may go through. The stage is identi
-5. **Label**: `Pending shepherd recommendation <https://github.com/ghc-proposals/ghc-proposals/pulls?q=is%3Aopen+is%3Apr+label%3A%22Pending+shepherd+recommendation%22>`_.  **Responsibility for next action:** shepherd.  **Timescale**: two weeks.
+6. **Label**: `Pending shepherd recommendation <https://github.com/ghc-proposals/ghc-proposals/pulls?q=is%3Aopen+is%3Apr+label%3A%22Pending+shepherd+recommendation%22>`_.  **Responsibility for next action:** shepherd.  **Timescale**: two weeks.
 
    The shepherd reviews the proposal, and may request changes that they would like to see before they submit their recommendation to accept or reject the proposal to the committee.
```
While we're here, the next two paragraphs (not actually part of the diff) could be clearer that if the shepherd requests revisions, they are not also expected to submit a recommendation within two weeks. Perhaps we should say "Within two weeks, if the shepherd does not request revisions, then they should..."

adamgundry commented on README.rst:
```diff
@@ -128,19 +174,30 @@ This section outlines what stages a proposal may go through. The stage is identi
-   may do that, or someone else. We mark the proposal as “implemented” once it
-   hits GHC’s ``master`` branch (and we are happy to be nudged to do so by
-   email, GitHub issue, or a comment on the relevant pull request).
+11. **Label**: `Dormant <https://github.com/ghc-proposals/ghc-proposals/pulls?q=is%3Aopen+is%3Apr+label%3A%22Dormant%22>`__.  If a proposal sees no activity for along time, it is marked as “dormant”,
+    and eventually closed.
```
I wonder if the Dormant state is actually helpful. There aren't clear guidelines about when it applies so we don't use it very consistently, neither do we usually close proposals if they are merely stalled due to lack of activity (because there are more good ideas than people having cycles to push them through the process). So I'm tempted to remove the Dormant state. What do others think?

VitWW replied:
I shifted Dormant between "Pending shepherd" and "Committee discussion" stages

adamgundry commented on README.rst:
```diff
@@ -128,19 +174,30 @@ This section outlines what stages a proposal may go through. The stage is identi
 
-7. **Label**: `Dormant <https://github.com/ghc-proposals/ghc-proposals/pulls?q=is%3Aopen+is%3Apr+label%3A%22Dormant%22>`__.  If a proposal sees no activity for along time, it is marked as “dormant”,
-   and eventually closed.
+8. **Label**: `Pending committee review <https://github.com/ghc-proposals/ghc-proposals/pulls?q=is%3Aopen+is%3Apr+label%3A%22Pending+committee+review%22>`__.  
```
I think this should be Accepted (but the PR is still open)? Then the next stage is Accepted + PR merged.

VitWW replied:
Almost.
I added "(No special label, but Accepted)" and "Accepted + Merged" respectively.

adamgundry commented on README.rst:
```diff
@@ -61,7 +61,47 @@ and language *stability*
++-----------------------------------+----------------+----------------------------------+--------------------------+----------------------------------+
+| Awaiting shepherd                 | 1 week         | Committee secretary              | Assign shepherd          | Awaiting shepherd recommendation |
++-----------------------------------+----------------+----------------------------------+--------------------------+----------------------------------+
+| Awaiting shepherd recommendation  | 2 weeks        | Shepherd                         | Make recommendation      | Committee discussion             |
```
```suggestion
| Pending shepherd recommendation   | 2 weeks        | Shepherd                         | Make recommendation      | Committee discussion             |
```

adamgundry commented on README.rst:
```diff
@@ -80,16 +120,22 @@ This section outlines what stages a proposal may go through. The stage is identi
    * `≡ List of proposals under discussion <https://github.com/ghc-proposals/ghc-proposals/pulls?q=is%3Aopen+is%3Apr+no%3Alabel>`__
  
-   At some point the proposal author brings the proposal before the committee for review, which will prompt the committee secretary to appoint a shepherd for the proposal.
+4. (No label.)  **Responsibility for next action:** author.
+
+   *Note: The Reviewing process does NOT start automatically when the discussion ends.*
+
+   At some point, the proposal author wishes to start a review process of the proposal by the committee. 
+
+   Then the author MUST explicitly ask in comments the committee secretary to bring the proposal before the committee for review (and to appoint a shepherd for it).
```
These are helpful clarifications, thanks. I think they belong as part of the previous "community discussion" stage, though? This isn't a separate stage, but explains how to transition to the next stage.

adamgundry commented on README.rst:
```diff
@@ -61,7 +61,47 @@ and language *stability*
+| Stage                             | Max duration   | Responsibility                   | Action                   | New stage                        |
+|                                   |                | for next action                  |                          |                                  |
++===================================+================+==================================+==========================+==================================+
+| Proposal creating                 | Indefinite     | Author                           | Pull request             | Community discussion             |
```
```suggestion
| Proposal creation                 | Indefinite     | Author                           | Pull request             | Community discussion             |
```

adamgundry commented on README.rst:
```diff
@@ -61,7 +61,47 @@ and language *stability*
+
+The stage is identified by a GitHub label, which is identified in the following list.
 
 1. (No label.) The author drafts a proposal.
```
I think we could merge points 1 and 2 so they correspond to the "proposal creation" stage in the table.

adamgundry commented on README.rst:
```diff
@@ -61,7 +61,47 @@ and language *stability*
+|                                   |                +----------------------------------+--------------------------+                                  |
+|                                   |                | Shepherd                         | Merge                    |                                  |
++-----------------------------------+----------------+----------------------------------+--------------------------+----------------------------------+
+| Accepted / Merged                 | Indefinite     | Anyone (including the author)    | Implement the proposal   | Implemented                      |
+|                                   |                | willing to implement it          |                          |                                  |
```
```suggestion
| Accepted / Merged                 | Indefinite     | Unspecified                      | Implement the proposal   | Implemented                      |
```

adamgundry commented on README.rst:
```diff
@@ -128,19 +174,30 @@ This section outlines what stages a proposal may go through. The stage is identi
+9. **Label**: `Accepted <https://github.com/ghc-proposals/ghc-proposals/pulls?q=label%3A%22Accepted%22>`__.  
+   **Responsibility for next action:** shepherd and committee secretary.  **Timescale**: one week. 
+
+   If the committee fully accepts the proposal, the committee secretary and/or shepherd marks the proposal as “accepted” and merges the Proposal.
```
```suggestion
   If the committee fully accepts the proposal, the committee secretary and/or shepherd marks the proposal as “accepted”, merges the PR and (if necessary) creates a tracking ticket on the GHC issue tracker.
```

adamgundry commented on README.rst:
```diff
@@ -128,19 +174,30 @@ This section outlines what stages a proposal may go through. The stage is identi
+  
+   If the committee accepts the proposal with minor corrections, the author and shepherd work together to make any final edits.
+
+9. **Label**: `Accepted <https://github.com/ghc-proposals/ghc-proposals/pulls?q=label%3A%22Accepted%22>`__.  
+   **Responsibility for next action:** shepherd and committee secretary.  **Timescale**: one week. 
+
+   If the committee fully accepts the proposal, the committee secretary and/or shepherd marks the proposal as “accepted” and merges the Proposal.
+   Any later changes must be submitted as a separate amendment proposal.
```
Let's add a link to the acceptance checklist (https://github.com/ghc-proposals/ghc-proposals/blob/master/acceptance.rst).

VitWW replied:
done.

adamgundry commented on README.rst:
```diff
@@ -128,19 +174,30 @@ This section outlines what stages a proposal may go through. The stage is identi
+   Any later changes must be submitted as a separate amendment proposal.
+
+10. **Label**: `Implemented <https://github.com/ghc-proposals/ghc-proposals/pulls?q=is%3Apr+label%3A%22Implemented%22>`__.   Once a proposal is accepted, it still has to be implemented.  The author
+    may do that, or someone else. We mark the proposal as “implemented” once it
```
I think it's worth being clear that acceptance of a proposal does not (unfortunately) guarantee someone will step up to implement it.

```suggestion
    may do that, or ask someone else to do so. GHC is a volunteer-driven project, so there is unfortunately no guarantee that accepted proposals will be implemented promptly.
    
    We mark the proposal as “implemented” once it
```

adamgundry reviewed (commented):
Thanks @VitWW, it's very helpful to clarify the process like this. I have some suggestions about details but am generally very much in favour.

VitWW commented:
Ok Adam, I added what you asked (see more details in "show resolved" on each task).

Should I change something else?

VitWW commented:
I also add from [issue 560](https://github.com/ghc-proposals/ghc-proposals/issues/560) reason why RST has preference over MD format.

maralorn commented:
Yeah, thank you. If I get super nitpicky I would say you should use "/" instead of "&" in the "Responsibility …" column since it is more an "or" than an "and" I think.

noughtmare commented on README.rst:
```diff
@@ -128,26 +198,44 @@ This section outlines what stages a proposal may go through. The stage is identi
 ---------------------------
 
-Proposals are written in either `ReStructuredText <http://www.sphinx-doc.org/en/stable/rest.html>`_ or `Markdown <https://github.github.com/gfm/>`_. While the proposal process itself has no preference, keep in mind that the `GHC Users Guide <http://downloads.haskell.org/~ghc/latest/docs/html/users_guide/editing-guide.html>`_ uses ReStructuredText exclusively. Accepted proposals written in ReStruc ...
+Proposals are written in `ReStructuredText <http://www.sphinx-doc.org/en/stable/rest.html>`_ . 
+
+Using `Markdown <https://github.github.com/gfm/>`_ instead is also allowed. But while the proposal process itself has no preference for formatting, 
+keep in mind that the `GHC Users Guide <http://downloads.haskell.org/~ghc/latest/docs/html/users_guide/editing-guide.html>`_ uses ReStructuredText exclusively. 
+Accepted proposals written in ReStructuredText thus have the slight benefit that they can be more easily included in the official GHC documentation. 
+In the other case, if the proposal is accepted, someone must volunteer (maybe the author themselves) to convert the Proposal into ReStructuredText format. 
+That's why the ReStructuredText format has a clear preference for the Committee to accept the proposal.
```
```suggestion
Proposals are written in `ReStructuredText <http://www.sphinx-doc.org/en/stable/rest.html>`_ . 
```

The conclusion from #560 was just to always ask for reStructuredText and remove the markdown option from the readme, so I think we can just remove all this text.

VitWW replied:
I add a claim, that if Shepherd agree, the Proposal need to be converted into RST before Committee starts.

>  If the proposal written in Markdown is approved by "Shepherd recommendation", Shepherd will ask the author to convert the Proposal into ReStructuredText format before the whole Committee starts to discuss.

I think it is a reasonable compromise between many different points of view in #560

VitWW replied:
Ok, I've removed mentioning of MD!
But what to do with "template.md"? Is I need to delete it?

VitWW commented on README.rst:
```diff
@@ -128,32 +198,46 @@ This section outlines what stages a proposal may go through. The stage is identi
 
-7. **Label**: `Dormant <https://github.com/ghc-proposals/ghc-proposals/pulls?q=is%3Aopen+is%3Apr+label%3A%22Dormant%22>`__.  If a proposal sees no activity for along time, it is marked as “dormant”,
-   and eventually closed.
+8. **Label**: (No special label, but Accepted).
```
done!

VitWW commented:
@simonpj All but (3) are done!

Is it better now?

simonpj commented:
I see that the stage column has links, but I expected the link for a particular stage to lead to a list of propsals at that stage or in that state.  But it doesn't.  It leads somwhere else.

Also could the sections have section numbers please?  That's an easy change to make.

VitWW commented:
> I see that the stage column has links, but I expected the link for a particular stage to lead to a list of proposals at that stage or in that state. But it doesn't. It leads somewhere else.

I'm sorry. Sometimes I see what I wish to see )))

1 done and 2 done!

simonpj commented:
Great.  

* Let's change "Acception" to "Finalisation" -- label "Accepted with PR still open".  This is when the PR is being finalised
* And change "Finalisation" to "Accepted" -- label "Accepted".  These are the accepted proposals.
* Accepted proposals are also called "Merged" and some points, which is confusing.
* Can we number the sections, please?
* Test all the links in the table.  I'm not sure they all work.

VitWW commented on README.rst:
```diff
@@ -61,46 +63,192 @@ and language *stability*
+|                                     | label               |                | | for next action      |                             |                              |
+|                                     |                     |                |                        |                             |                              |
++=====================================+=====================+================+========================+=============================+==============================+
+| |1|)    Proposal creation           | -                   | Indefinite     | Author                 | Pull request                | Community discussion         |
```
Do you prefer to render as `-`? Ok, done!

adamgundry commented:
@VitWW thanks for continuing to make progress on this. I've already emailed the committee to draw attention to this PR, but I don't think it needs the full committee review process. I'd like to polish it a bit further with some final edits, then we can merge. (I've simplified the table a little on a branch (https://github.com/adamgundry/ghc-proposals/blob/update-readme/README.rst), but I haven't yet had time to edit the detailed list of stages.)

@simonpj are you happy with the removal of the Dormant state? I don't think it adds much, and I'd like to make this as simple as possible. (Assuming we agree I will go ahead and remove the label altogether.)

simonpj commented:
> My suggestion is that we scrap the "Dormant" label and instead close PRs where it is clear that the author has abandoned the effort and there is unlikely to be interest in pursuing the idea from anyone else.

This is exactly what I think Dormant is for.  I suppose closing would be an alternative, but
* Can one search for closed proposals as easily as for Dormant ones?
* If a proposal is accepted an implement, does it never get closed?  (We would not want to mix them up with Dormant.) 
* A proposal might be closed just because the author made a mistake and opened two, or becuase they became convinced it was an outright bad idea.

My instinct is to maintain a distinction between "closed" and "Dormant"; and perhaps to more assiduously move domant proposals into the Dormant state -- e.g. after no discussion for 3 months.  I can always be re-awakened.

VitWW commented:
@adamgundry Ok, I've updated the "Stages" table and added "outcome" table too.
I see your vision of links in the table are similar to Graf's vision. Done.
Links from stages names leads to detailed description, links from "GitHub label" leads to filtered PRs with same labels.
I hope, Simon PJ agree with such change.

I also agree to split "Community discussion" stage into 2 stages: "Community discussion" and "Needs revision".
But I don't know how to fill this stage at detailed description.
So, help me please to fill.
Right now I wrote:

> The wider community discusses the proposal in the commit section of the pull request, while the author refines the proposal.
> But the author mainly focuses on revising the proposal via the guidance of the Shepherd and the Committee's view.

I see, you are agree with Graf to remove "Continue discussion" action from the table. Done! But I remain this action in description on "Community discussion" stage. Do I need to remove it?

I also see you wish to unite "Finalisation" and "Accepted" stages, but unfortunately 
- those stages has different responsibilities and labels and it make sense to have 2 stages
- Simon PJ wish to have different labels
So, I remain those stages separated. At least until you resolve this difference with Simon.

@simonpj I hope I added everything you asked about. Do you wish anything else to change?

@sgraf812 I did what you asked (see more details in "show resolved" on each task). Do you wish anything else to change?

VitWW commented:
Ok, Simon wish to have "Dormant" label. I added it to tables and descriptions (2nd and 5th stages).

Is it better now?

simonpj commented:
> If a proposal is accepted an implemented, does it never get closed? (We would not want to mix them up with Dormant.)

Aha. I see that all accepted proposals are in fact closed!  (See here.)  So if we don't have "Domant" you'd have to search for "Closed-and-no-label" if indeed it is possible to do that.  Easier just to have Dormant.

adamgundry commented:
> Aha. I see that all accepted proposals are in fact closed! (See here.) So if we don't have "Domant" you'd have to search for "Closed-and-no-label" if indeed it is possible to do that. Easier just to have Dormant.

One can indeed search for closed PRs with or without labels. But I think the mechanics of labels are not that crucial here - as suggested in https://github.com/ghc-proposals/ghc-proposals/issues/706#issuecomment-3194355132 we may want to use a project board instead (in which case it would make sense to have a column for Dormant proposals, and it would be easier to keep it up to date).

On reflection, I think it could make sense for the main list of open PRs to be those in an "active" stage, which implies that we should close PRs considered to be Dormant (just as Accepted and Rejected PRs are closed as there is no activity on the committee side).

Then looking at the latest version of the table, is there a useful difference between Dormant and Withdrawn? Can we just say that there is one stage that covers both cases?

VitWW commented:
Adam, I see you don't like trailing spaces )))

>  I did a final pass including adding text along the lines you suggest.

**Yes**, right now text looks much nicer!

And a bit of perfection - I changed the order in the table and make first choice - "good choice" ("Submitting")

I also added to `0000-template.rst`
```
Acknowledgements
----------------

(Optional) This section provides an opportunity to say thanks
to third parties for their contributions to the proposal.
```
If you against, be free to delete/revert this change.

VitWW commented:
Improvements were made to mostly README file about proposal process, following propositions was discussed at #706 and #560 issues.
@adamgundry (and me too) thinks that current view is a final view and it is ready to merge.

@simonpj , @maralorn,  @sgraf812 and everyone whom those changes are interested in - are you happy with changes?

Summary in brief of changes was made:
- table of Stages (and Outcomes) was added for easier understanding the whole machinery
- detailed description list was synced with the table
- detailed description text was made more clear and explicit
- links was added from the table to Stages descriptions (and vice versa)
- links was added from the table to filtered proposals
- accepted steps was synced with the table
- "Dormant" label  became outdated
- MarkDown format became forbidden and file "0000-template.md" was removed
- links to "guidelines for respectful communication" was added to discussion parts
- explanation was added how to propose changes in comments to PR
- optional "Acknowledgments" was added to template
- notification to whole committee for early review at the beginning of Shepherd review Stage was added
- optional additional notification from Secretary to author was added to explain the author's responsibility to ask for committee review
- author responsibility to edit and correct the proposal was described at "Needs review" and "Accepted" Stages
